### PR TITLE
cataclysmDDA: add modding interface

### PR DESCRIPTION
**Pull request by mnacamura**
###### Motivation for this change
This PR adds a basic interface to package Cataclysm DDA mods, soundpacks, and tilesets. See [this](/NixOS/nixpkgs/pull/84261#issuecomment-609071853) and [this](/NixOS/nixpkgs/pull/84261#issuecomment-609604399) comments. It also adds UndeadPeople tileset, which was dropped in the latest version bump.

Minor improvements:
- Make `cataclysm-dda-git` to be overidden correctly
- Support build of `cataclysmdda{,-git}` with `USE_XDG_DIR` flag
- Apply the troublesome locale patch dynamically

~~This PR should be merged after NixOS/nixpkgs/pull/84261.~~ DONE

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by svenkeidel — APPROVED** [review]

**Comment by mnacamura** [comment]
On Darwin, the launcher fails to find installed mods. I'll try to fix it.

**Comment by mnacamura** [comment]
Fixed the launcher problem. Confirmed that it builds and runs successfully on NixOS and Darwin.
@svenkeidel @catern Could you kindly review it (again)?

**Review by bbigras — APPROVED** [review]

**Comment by bbigras** [comment]
I didn't check all the nix files but it builds and runs fine for me. :+1: 

I also tested the launcher on NixOS with gnome and using the undead people tileset.

Thank you very much for this PR.

**Review by svenkeidel — APPROVED** [review]

File: pkgs/games/cataclysm-dda/wrapper.nix [hunk 1]
```diff
@@ -28,5 +28,21 @@ else symlinkJoin {
         wrapProgram $out/bin/cataclysm-tiles \
             --add-flags "--datadir $out/share/cataclysm-dda/"
     fi
+
+    # Launch the wrapped proglam
```
Comment (worldofpeace) [inline pkgs/games/cataclysm-dda/wrapper.nix:32]:
```suggestion
    # Launch the wrapped program
```

you need to correct the commit message also

Comment (mnacamura) [reply pkgs/games/cataclysm-dda/wrapper.nix]: Sorry. Fixed.

**Comment by svenkeidel** [comment]
I tested this PR multiple times and it works perfectly. Can we merge this?

**Review by svenkeidel — APPROVED** [review]

**Comment by mnacamura** [comment]
ping @worldofpeace

**Comment by worldofpeace** [comment]
There's one glaring issue with this PR, and it's the documentation. It's not a part of any user facing material like the nixpkgs manual. I believe that section would belong there because it's "how to configure a package". We luckily can include markdown into the nixpkgs manual, if it was the nixos one you'd have to translate it to docbook. I don't think our documentation on how to do this is sufficient however https://nixos.org/nixpkgs/manual/#chap-contributing (and I'm not sure how either). Perhaps asking how in #nixos-dev will help.

**Comment by worldofpeace** [comment]
My best guess is the trail left by the python framework:

In the `launguage-framework/index.xml` it includes the python section like
https://github.com/NixOS/nixpkgs/blob/70450f1d5ec8d4f7a925d100f69b0176d22d08fd/doc/languages-frameworks/index.xml#L26

my intuition tells me pandoc converts the section into a file name like this?

In that same directory, the actual file is named `python.section.md`
https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md

I'm not sure that will be helpful, but it's the best can intuit at the moment.

**Comment by mnacamura** [comment]
@worldofpeace Many thanks for your help. I moved the document to nixpkgs manual.

**Comment by mnacamura** [comment]
There is another version bump for the stable version: /pull/93349. I will rebase this PR after that has been merged.

**Comment by mnacamura** [comment]
Rebase done.

**Comment by svenkeidel** [comment]
Hi @mnacamura,

I just tried f2b2347 and got
```
nix-build -A 'cataclysm-dda-git.withMods (mods: with mods; [ tileset.UndeadPeople ])'
error: attribute 'withMods (mods: with mods; [ tileset' in selection path 'cataclysm-dda-git.withMods (mods: with mods; [ tileset.UndeadPeople ])' not found
```
Any ideas?

**Comment by svenkeidel** [comment]
The documentation looks good though :+1:

**Comment by mnacamura** [comment]
@svenkeidel Thanks for checking the document. I think `nix-build`'s `-A` flag accepts only attribute path, not expression. Try `nix-build -E 'with import <nixpkgs> {}; cataclysm-dda-git.withMods (mods: with mods; [ tileset.UndeadPeople ])'`

**Comment by svenkeidel** [comment]
:man_facepalming: thanks. Sometimes the nix tooling can be confusing.

**Comment by worldofpeace** [comment]
Thank you @mnacamura

**Pull request merged**